e2e test setup

asalkeld opened the pull request:
This creates and tears down the vnet and cluster.
Then does a simple scale up/down

asalkeld committed:
run dev ensure

asalkeld committed:
begining of e2e test

asalkeld committed:
fixes to get it working

asalkeld committed:
Very simple scale up/down test with very little verifcation

asalkeld committed:
add a check for Microsoft.Resources/deployments/write

note: just retrying didn't help

jim-minter commented on test/e2e/setup.go:
```diff
@@ -0,0 +1,367 @@
+		logger.Error(err)
+	}
+	logger.Infof("AfterSuite deleting vnet %s", cs.rg)
+	err = cs.deployments.DeleteAndWait(ctx, cs.rg, "azuredeploy")
```
I don't believe that deleting a deployment deletes the things it deployed.  Just use a vnet client for the creation and deletion.

jim-minter commented on test/e2e/scaleupdown.go:
```diff
@@ -0,0 +1,59 @@
+)
+
+var _ = Describe("Scale Up/Down E2E tests [ScaleUpDown][EveryPR][LongRunning]", func() {
+	openshiftclusters := redhatopenshift.NewOpenShiftClustersClientWithBaseURI("https://localhost:8443", os.Getenv("AZURE_SUBSCRIPTION_ID"))
```
the az aro extension checks if RP_MODE==development to decide whether to override the base URI + InsecureSkipVerify or not - please do the same.

jim-minter replied:
do we want the client setups all together, rather than per-test?

jim-minter commented on test/e2e/scaleupdown.go:
```diff
@@ -0,0 +1,59 @@
+	scale := func(count int32) {
+		ctx := context.Background()
+		By("Fetching the manifest")
+		external, err := openshiftclusters.Get(ctx, os.Getenv("CLUSTER"), os.Getenv("CLUSTER"))
```
at the moment prod doesn't support the case where cluster RG == cluster name.  Please pick them so they are different.

jim-minter commented on test/e2e/scaleupdown.go:
```diff
@@ -0,0 +1,59 @@
+
+	It("should be possible to maintain a healthy cluster after scaling it out and in", func() {
+		By("Scaling up")
+		scale(2)
```
I'm confused - are you scaling down from 3 to 2 here?

jim-minter commented on test/e2e/scaleupdown.go:
```diff
@@ -0,0 +1,59 @@
+		scale(2)
+
+		By("Scaling down")
+		scale(1)
```
and then to 1?  Does the RP even allow that?  It should be erroring - can you check please?

jim-minter commented on test/e2e/scaleupdown.go:
```diff
@@ -0,0 +1,59 @@
+		Expect(err).NotTo(HaveOccurred())
+
+		By("waiting for cluster to scale")
+		err = future.WaitForCompletionRef(ctx, openshiftclusters.Client)
```
add a wrapper around the client to avoid having to call WaitForCompletionRef

jim-minter commented on test/e2e/scaleupdown.go:
```diff
@@ -0,0 +1,59 @@
+
+var _ = Describe("Scale Up/Down E2E tests [ScaleUpDown][EveryPR][LongRunning]", func() {
+	openshiftclusters := redhatopenshift.NewOpenShiftClustersClientWithBaseURI("https://localhost:8443", os.Getenv("AZURE_SUBSCRIPTION_ID"))
+	openshiftclusters.PollingDuration = time.Minute * 90
```
60 should be fine - put in a client wrapper

jim-minter commented on test/e2e/scaleupdown.go:
```diff
@@ -0,0 +1,59 @@
+})
+
+func setCount(oc *redhatopenshift.OpenShiftCluster, count int32) error {
+	for _, p := range *oc.Properties.WorkerProfiles {
```
we have a lot of code that hard codes WorkerProfiles[0] - this can do the same with a // TODO

jim-minter commented on test/e2e/setup.go:
```diff
@@ -0,0 +1,367 @@
+var (
+	kubeadminPassword string
+	apiVersions       = map[string]string{
+		"network": "2019-07-01", // copied from 0-installstorage
```
use the vnet client, then you won't need this

jim-minter commented on test/e2e/setup.go:
```diff
@@ -0,0 +1,367 @@
+		return err
+	}
+	c.log.Infof("waiting for successful provision of resource group %s", c.rg)
+	return wait.PollImmediate(5*time.Second, 5*time.Minute, func() (bool, error) {
```
resource group creation is synchronous - this is not needed

jim-minter commented on test/e2e/setup.go:
```diff
@@ -0,0 +1,367 @@
+}
+
+func (c *clientSet) ensureResourceGroup(ctx context.Context) error {
+	if ready, _ := c.checkResourceGroupIsReady(ctx); ready {
```
CreateOrUpdate is idempotent - shouldn't need checkResourceGroupIsReady

jim-minter commented on test/e2e/setup.go:
```diff
@@ -0,0 +1,367 @@
+
+func randomAddressPrefix() (string, error) {
+	//	10.$((RANDOM & 127)).$((RANDOM & 255)).0/24
+	a, err := rand.Int(rand.Reader, big.NewInt(int64(127)))
```
slightly overkill - math/rand is fine for this

jim-minter commented on test/e2e/setup.go:
```diff
@@ -0,0 +1,367 @@
+		return err
+	}
+	c.log.Debugf("assigning role to AZURE_FP_CLIENT_ID for vnet")
+	err = c.assignVnetRoleTo(ctx, os.Getenv("AZURE_FP_CLIENT_ID"))
```
need to choose between AZURE_FP_CLIENT_ID and "f1dd0a37-89c6-4e07-bcd1-ffd3d43d8875" depending on RP_MODE - see custom.py

jim-minter commented on test/e2e/setup.go:
```diff
@@ -0,0 +1,367 @@
+}
+
+func (c *clientSet) createCluster(ctx context.Context, vnetRG, clusterName string) error {
+	b, err := ioutil.ReadFile("cluster.json")
```
I got rid of examples/ in a concurrent commit - can we try not using a json file here?

jim-minter commented on test/e2e/setup.go:
```diff
@@ -0,0 +1,367 @@
+
+var _ = AfterSuite(func() {
+	logger := utillog.GetTestLogger()
+	cs, err := newClientSet(
```
wonder about making this a global

jim-minter commented on test/e2e/setup.go:
```diff
@@ -0,0 +1,367 @@
+	logger := utillog.GetTestLogger()
+	for _, key := range []string{
+		"LOCATION", "AZURE_SUBSCRIPTION_ID", "AZURE_TENANT_ID",
+		"CLUSTER", "VNET_RESOURCEGROUP",
```
Maybe e2e should pick its own CLUSTER and RESOURCEGROUP and use RESOURCEGROUP instead of VNET_RESOURCEGROUP

jim-minter commented on test/e2e/setup.go:
```diff
@@ -0,0 +1,367 @@
+	for _, key := range []string{
+		"LOCATION", "AZURE_SUBSCRIPTION_ID", "AZURE_TENANT_ID",
+		"CLUSTER", "VNET_RESOURCEGROUP",
+		"AZURE_ARM_CLIENT_ID", "AZURE_FP_CLIENT_ID", "AZURE_CLUSTER_CLIENT_ID",
```
at least AZURE_ARM_CLIENT_ID should not be needed

jim-minter commented on test/util/log/log.go:
```diff
@@ -0,0 +1,19 @@
+
+func GetTestLogger() *logrus.Entry {
+	logrus.SetLevel(logrus.DebugLevel)
+	logrus.SetOutput(ginkgo.GinkgoWriter)
```
doesn't this entirely duplicate GetLogger() with the exception of this line?

jim-minter commented on pkg/util/azureclient/resources/deployments_addons.go:
```diff
@@ -19,3 +20,12 @@ func (c *deploymentsClient) CreateOrUpdateAndWait(ctx context.Context, resourceG
 	return future.WaitForCompletionRef(ctx, c.DeploymentsClient.Client)
 }
+
+func (c *deploymentsClient) DeleteAndWait(ctx context.Context, resourceGroupName string, deploymentName string) error {
```
think this can be reverted

asalkeld committed:
rebase and respond to review comments

Major changes

- examples/cluster-v20191231.json has gone, use a struct directly
- fix the scale number
- make clients global
- use vnet client not a deployment
- create an openshiftclusters wrapper client

asalkeld committed:
fix the resourceGroup in the frontend

This was using the clusterName and ignoring the resourceGroupName in the resourceID

jim-minter commented:
Closing for https://github.com/Azure/ARO-RP/pull/24
